Make derivative calculations work with xarray DataArrays

jthielen opened the pull request:
While this was originally devised while working on isallobaric wind and thermal wind, I've also found that it would be needed for some of the cross-section examples I'm putting together (a couple forms of PV that our current implementations don't handle), so here it is now on its own.

This PR modifies the current derivative calculations to work with DataArrays themselves, rather than by conversion to (and output as) pint quantities. This includes automatic coordinate handling. With that, in order to support lat/lon non-uniform coordinates, the `lat_lon_grid_deltas` function also needed to be moved from kinematics.py to tools.py to make the imports work right.

Three implementation questions I'd like to ask right away:
- what should the default be for `first_derivative` and `second_derivative` if `axis` is not given, or should it be required (what I have right now)?
- would it be useful to add a "with respect to" option (`wrt`, following CF?) to automatically pick out the desired axis? For example, if we wanted the rate of change of temperature with respect to time, we could just have
  ```
  first_derivative(temperature, wrt='time')
  ```
  vs. needing to know the name of the time axis,
  ```
  first_derivative(temperature, axis='time1')
  ```
  or
  ```
  first_derivative(temperature, axis=temperature.metpy.time.name)
  ```
- should the old v0.8 preprocessing approach still be made to work here, but be deprecated?

dopplershift commented:
* Default should stick with `axis=0`
* Let's go with:
  1. Check type; if int, just use it
  2. If string, first see if it matches a dimension, if so use it
  3. Otherwise, see if we can match it using our aliases (e.g. 'time')
* Keep existing behavior working. If user gives us `x` or `dx`, use them and convert any `DataArray` to `pint.Quantity`.

jthielen commented:
I made the discussed changes, and added corresponding tests, but I also went a little past that and changed how I did some of the implementation. After playing around with a few iterations, I added in a `grid_deltas_from_dataarray` function in `calc` and a `axis_name` method in the xarray accessor... feedback on those names would be appreciated. I think what I have now for the derivative wrapper is a little bit cleaner, and the `grid_deltas_from_dataarray` could be useful in general as well (it reduces some boilerplate code in the current state of working with kinematics calculations and DataArrays).

dopplershift commented on metpy/xarray.py:
```diff
@@ -104,6 +108,39 @@ def coordinates_identical(self, other):
         return True
 
+    @property
+    def as_seconds(self):
```
How about `as_timestamp()` to match the standard library functions?

dopplershift commented on metpy/xarray.py:
```diff
@@ -104,6 +108,39 @@ def coordinates_identical(self, other):
+                            dims=self._data_array.dims,
+                            attrs=attrs)
+
+    def axis_name(self, axis):
```
`axis_from_name`?

dopplershift replied:
No. :sheep:

`find_axis_name`? I think I want a verb in there some how, since as a noun it's really more of a property, which isn't the case.

jthielen commented on metpy/calc/tools.py:
```diff
@@ -772,24 +773,212 @@ def _less_or_close(a, value, **kwargs):
+    """
+    if f.metpy.crs['grid_mapping_name'] == 'latitude_longitude':
+        dx, dy = lat_lon_grid_deltas(f.metpy.x.values, f.metpy.y.values,
+                                     initstring=f.metpy.cartopy_crs.proj4_init)
```
Yes, since we have the information, I thought we mind as well use it to get a more correct proj4 geod.
